tracing-subscriber/src/fmt/writer.rs

```diff
@@ -510,7 +510,9 @@ pub struct MutexGuardWriter<'a, W>(MutexGuard<'a, W>);
 /// [`print!`]: std::print!
 #[derive(Default, Debug)]
 pub struct TestWriter {
-    _p: (),
+    /// Whether or not to use `stderr` instead of the default `stdout` as
+    /// the underlying stream to write to.
+    use_stderr: bool,
 }
```

Comment:
Would it be reasonable here to implement this with a generic writer instead of a bool? E.g.:
```rust
pub struct TestWriter<W: Write = Stdout> {
    writer: W
}

impl Default for TestWriter<Stdout> { ... }

impl TestWriter<Stderr> {
    fn with_stderr() { ... }
}
```

I don't think this causes a breaking change, but I'm not 100% certain of this.

Reply:
Also not sure; but breakage seems extremely unlikely.

In any event, I don't believe this accomplishes the same thing. As per my understanding we have to use `print` / `eprint` macros to hook into the test output capture logic. I believe usage of `Stdout` circumvents that and would result in unconditionally visible output.

Reply:
```rust
write!(self.writer, ...)
```
Would be the equivalent there. the print / eprint macros under the hood just write to stdout/stderr (afaik).

Reply:
I suppose we could add the `Writer` generic and then just ignore it and `print` / `eprint` for the time being. That should be forward compatible.

Reply:
On the other hand, the entire concept of `TestWriter` is only a thing *because* of this limitation, I believe. Without it, you could just use `with_writer`, unless I am missing something.

Reply:
> On the other hand, the entire concept of `TestWriter` is only a thing _because_ of this limitation, I believe. Without it, you could just use `with_writer`, unless I am missing something.

I missed that. Yup you're right there too. If libtest is fixed, then eventually deprecating `TestWriter` would make sense. If there's an enum in a reasonable place that hads Stdout and Stderr variants (e.g. in std::io or an existing dep on tracing - I don't think this exists) , then that enum might be a marginally better value over the bool, but it's no big deal since this makes sense regardless and is private. It could be worth linking to that issue mentioned above in the comment to avoid the same conversation next time someone looks at this. But otherwise this LGTM.

Reply:
> It could be worth linking to that issue mentioned above in the comment to avoid the same conversation next time someone looks at this.

Happy to add, I am just not exactly sure what you have in mind, to be honest. Can you perhaps add a suggestion?

Reply:
Done. Marking this comment as resolved.